### PR TITLE
CAL-154 Automates the Sample NSILI Client tests cases into the itests

### PR DESCRIPTION
**Pull request by emmberk**
#### What does this PR do?

This PR automates the Sample NSILI Client tests cases into the itests to expand coverage.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?

@lcrosenbu @bdeining @troymohl @glenhein @mweser @codice/test 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).

@kcwire
@lessarderic

#### How should this be tested?

1. Confirm itests pass on full build.
2. Confirm that the Sample NSILI Client tests cases are still available to be run from the command line (see [README](https://github.com/codice/alliance/blob/master/distribution/sdk/sample-nsili-client/README.md)). The functionality of the Sample NSILI Client tests cases should be the same as before. NOTE: The NSILI App must be installed for the NSILI Endpoint to be active.

#### Any background context you want to provide?

The goal of the Sample NSILI Client is to test that a client can make queries and orders to the NSILI Endpoint.

#### What are the relevant tickets?

[CAL-154](https://codice.atlassian.net/browse/CAL-154)

#### Checklist:
- [x] Documentation Updated
	- [ ] Change Log Updated
- [ ] Update / Add Unit Tests
- [x] Update / Add Integration Tests

### CODE REVIEW

File: catalog/nsili/nsili-app/src/main/resources/features.xml [hunk 1]
```diff
@@ -43,7 +43,7 @@
     </feature>
 
     <feature name="nsili-app" install="auto" version="${project.version}"
-             description="The Alliance NSILI App provides interoperability with STANAG 4559 compliant services. ::Alliance NSILI">
+             description="The Alliance NSILI App provides interoperability with STANAG 4559 compliant services.">
```
Comment (bdeining) [inline catalog/nsili/nsili-app/src/main/resources/features.xml]: Why this change?  I think `::Alliance NSILI` puts the name on the app in the Admin UI

File: distribution/sdk/sample-nsili-client/pom.xml [hunk 2]
```diff
@@ -60,19 +65,64 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.antlr</groupId>
+            <artifactId>antlr4-runtime</artifactId>
+            <version>4.5.1-1</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.4</version>
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-client/pom.xml]: ${commons-lang3.version}

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java [hunk 3]
```diff
@@ -13,121 +13,102 @@
...
     private static final Logger LOGGER = LoggerFactory.getLogger(Client.class);
 
-    private static final boolean SHOULD_PROCESS_PKG_ELEMENTS = false;
+    private static final int LISTEN_PORT = 8200;
 
-    private static final boolean SHOULD_TEST_STANDING_QUERY_MGR = false;
+    private static final String NSILI_FILE_URI_PATH = "/nsili/file";
 
-    private static final boolean SHOULD_DOWNLOAD_PRODUCT = true;
+//    private static final boolean SHOULD_PROCESS_PKG_ELEMENTS = true;
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java]: This comment can be removed

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/NsiliWeb.java [hunk 4]
```diff
@@ -1,72 +0,0 @@
...
-import javax.ws.rs.PathParam;
-import javax.ws.rs.Produces;
-import javax.ws.rs.core.MediaType;
-import javax.ws.rs.core.Response;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-@Path("/nsili")
-public class NsiliWeb {
```
Comment (emmberk) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/NsiliWeb.java]: Replaced with Restito StubServer

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java [hunk 5]
```diff
@@ -13,121 +13,102 @@
...
 
 public class Client {
-    private HttpServer server = null;
-
-    public static int LISTEN_PORT = 8200;
 
     private static final Logger LOGGER = LoggerFactory.getLogger(Client.class);
 
-    private static final boolean SHOULD_PROCESS_PKG_ELEMENTS = false;
+    private static final int LISTEN_PORT = 8200;
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java]: Is it worth it to make this configurable via `-Dexec.args`?

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java [hunk 6]
```diff
@@ -13,121 +13,102 @@
...
+                    } catch (Exception e) {
+                        LOGGER.info("Unable to test ProductMgr: {}",
+                                NsilCorbaExceptionUtil.getExceptionDetails(e));
                     }
+                    LOGGER.info("-----------------------");
+
+                    // OrderMgr
+                    PackageElement[] packageElements = sampleNsiliClient.order(results[i]);
+
+                    // TODO https://jira.di2e.net/browse/GDIB-1751
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java]: This link should be replaced with a DDF jira ticket

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+import com.xebialabs.restito.semantics.Call;
+import com.xebialabs.restito.semantics.Condition;
+import com.xebialabs.restito.server.StubServer;
+
+@RunWith(PaxExam.class)
+@ExamReactorStrategy(PerClass.class)
+public class NsiliEndpointTest extends AbstractAllianceIntegrationTest {
+
+    private static final String[] REQUIRED_APPS =
+            {"catalog-app", "solr-app", "spatial-app", "imaging-app", "nsili-app", "alliance-app"};
```
Comment (bdeining) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: Are `imaging-app` or `alliance-app` required to test the NSILI endpoint?

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        // TODO test packageElements of the order response https://jira.di2e.net/browse/GDIB-1751
+        assertStubServerPutRequest(orderResultFilename, "application/x-tar");
+
+        deleteMetacard(ingestedProductId);
+        sampleNsiliClient.cleanup();
+    }
+
+    @Test
+    public void testCatalogMgrViaCallback() throws Exception {
+        SampleNsiliClient sampleNsiliClient = new SampleNsiliClient(Integer.parseInt(RESTITO_STUB_SERVER_PORT.getPort()), NSILI_ENDPOINT_IOR_URL.getUrl());
```
Comment (bdeining) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: Does the sampleNsiliClient need to be started new for each test, or can it be moved into an @Before?

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    private void startHttpListener() {
+        server = new StubServer(Integer.parseInt(RESTITO_STUB_SERVER_PORT.getPort())).run();
+        whenHttp(server).match(method(Method.PUT), Condition.startsWithUri(NSILI_FILE_URI_PATH))
+                .then(Action.success());
+    }
+
+    private void assertStubServerPutRequest(String expectedFilename, String expectedType) {
+        assertThat(server.getCalls().size(), is(1));
+        final Call call = server.getCalls().get(0);
+        assertThat(call.getMethod(), is(equalTo(Method.PUT)));
```
Comment (bdeining) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: `equalTo` can be removed from these assertions

File: catalog/nsili/catalog-nsili-orb-impl/src/main/resources/OSGI-INF/metatype/metatype.xml [hunk 10]
```diff
@@ -17,7 +17,7 @@
     <OCD name="NSILI Corba ORB" id="org.codice.alliance.nsili.orb.impl.corbaorb"
          description="NSILI Corba ORB Configuration">
 
-        <AD description="Corba Listen Port."
+        <AD description="Corba Listen Port. NOTE: A restart is required to set the port back to a port number that has already been set."
```
Comment (harrison-tarr) [inline catalog/nsili/catalog-nsili-orb-impl/src/main/resources/OSGI-INF/metatype/metatype.xml]: What happens if you wait a few minutes between each set? I think that should allow you to then set the same port, without triggering that error. If that's not the case, it could be that object (the ORB?) is not correctly releasing it's resources, since I would think that once its port has changed, it should close and release that socket.

Comment (emmberk) [inline catalog/nsili/catalog-nsili-orb-impl/src/main/resources/OSGI-INF/metatype/metatype.xml]: After talking with @troymohl, it looks like the orb doesn't quite ever give up the post until a restart and there doesn't seem to be a satisfactory solution for this.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java (hunk 5 above)
Comment (emmberk) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java]: I don't think so. The only time that it's ever run with that port is from the command line by developers, and if you're running it there, you'll probably also be changing `SHOULD_PROCESS_PKG_ELEMENTS`, `SHOULD_TEST_STANDING_QUERY_MGR`, and `SHOULD_DOWNLOAD_PRODUCT`. I think simpler is better in this case, but I'm not set on that.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java (hunk 3 above)
Comment (emmberk) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java]: This boolean is used in the commented-out code in the runTest method that should be fixed in [CAL-192](https://codice.atlassian.net/browse/CAL-192).

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java (hunk 8 above)
Comment (emmberk) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: After wrestling the with itest dependencies, I decided that making a new sample client for each test method was the best solution for now.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 11]
```diff
@@ -270,42 +292,42 @@ public int getHitCount(Query query) throws Exception {
...
-    public void retrieveProductFromDAG(DAG dag) throws MalformedURLException {
+    public void downloadProductFromDAG(DAG dag) throws MalformedURLException {
         LOGGER.info("Downloading products...");
         for (int i = 0; i < dag.nodes.length; i++) {
             Node node = dag.nodes[i];
-            if (node.attribute_name.equals("productUrl")) {
+            if (node.attribute_name.equals(NsiliConstants.PRODUCT_URL)) {
 
                 String url = node.value.extract_string();
+                url = url.replaceAll("\\s+", "%20");
```
Comment (troymohl) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: This may be a bug in the code that sets the URL if it is leaving in spaces vs properly encoding the URL. Also, we should probably be using http://docs.oracle.com/javase/8/docs/api/java/net/URLEncoder.html instead of encoding it ourselves.

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java (hunk 7 above)
Comment (emmberk) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: Removed `imaging-app` and `alliance-app`

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java (hunk 5 above)
Comment (bdeining) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java]: 👍

File: distribution/sdk/sample-nsili-client/pom.xml [hunk 12]
```diff
@@ -60,19 +65,64 @@
...
-            <artifactId>jersey-core</artifactId>
-            <version>1.19</version>
+            <groupId>ddf.catalog.core</groupId>
+            <artifactId>versioning-common</artifactId>
+            <version>${ddf.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>joda-time</groupId>
+            <artifactId>joda-time</artifactId>
+            <version>2.3</version>
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-client/pom.xml]: ${joda-time.version}

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java (hunk 11 above)
Comment (emmberk) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: Replaced with `getEncodedUrlFromString` method

**Comment by emmberk** [comment]
Updated to use `NsiliConstants` and to not test results that have been deleted from the Catalog.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 13]
```diff
@@ -442,6 +478,7 @@ public void testStandingQueryMgr(POA poa, Query query) throws Exception {
             start.at(ResultDAGConverter.getAbsTime(startDate));
 
             LifeEvent end = new LifeEvent();
+            final long ONE_YEAR = 365L * 24L * 60L * 60L * 1000L;
```
Comment (glenhein) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: You don't need to change it, but for future reference, you can do: TimeUnit.DAYS.toMillis(365).

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java [hunk 14]
```diff
@@ -13,121 +13,111 @@
...
+
+                        // OrderMgr
+                        PackageElement[] packageElements = sampleNsiliClient.order(results[i]);
+
+                        // ProductMgr
+                        if (SHOULD_PROCESS_PKG_ELEMENTS) {
+                            // TODO prod is null for packageElement https://codice.atlassian.net/browse/CAL-192
+                            for (PackageElement packageElement : packageElements) {
+                                Product product = packageElement.prod;
+                                sampleNsiliClient.get_parameters(product);
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java]: Shouldn't we be doing something with the returned values, i.e., print or validate them?

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java [hunk 15]
```diff
@@ -142,23 +132,15 @@ public static void main(String args[]) {
             LOGGER.info("Unable to obtain IOR File :  Must specify URL to IOR file.");
         }
         try {
-            client.runTest(args);
+            client.runTest(args[0]);
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java]: Should we have a JavaDoc comment on `main` or on the class that explains what the expected argument is?

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 16]
```diff
@@ -131,54 +135,70 @@
...
-        this.poa = poa;
+    private void initPoa() throws InvalidName, AdapterInactive {
+        POA rootPOA = POAHelper.narrow(orb.resolve_initial_references("RootPOA"));
+        rootPOA.the_POAManager()
+                .activate();
+        poa = rootPOA;
     }
 
-    public void initLibrary(String iorFilePath) {
+    public void initCorbaLibrary(String iorUrl) throws Exception {
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: `private`?

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 17]
```diff
@@ -192,8 +212,10 @@ public void initLibrary(String iorFilePath) {
         return types;
     }
 
-    public void initManagers(String[] managers) throws Exception {
-        for (String managerType : managers) {
+    public void initManagers() throws Exception {
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: `private`?

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 18]
```diff
@@ -270,65 +292,84 @@ public int getHitCount(Query query) throws Exception {
...
-            printDAGAttributes(results[i]);
-            if (downloadProduct) {
-                try {
-                    retrieveProductFromDAG(results[i]);
-                } catch (MalformedURLException e) {
-                    LOGGER.error("Invalid URL used for product retrieval.", e);
-                }
-            }
+    private void printDagAttributes(DAG[] dags) {
+        for (DAG dag : dags) {
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: Could also do `dags.forEach(this::printDagAttributes)`.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 19]
```diff
@@ -270,65 +292,84 @@ public int getHitCount(Query query) throws Exception {
...
+     *
+     * @param dag
+     * @return hashmap of attributes and their values
+     */
+    public void printDagAttributes(DAG dag) {
         LOGGER.info("--------------------");
         LOGGER.info("PRINTING DAG ATTRIBUTES");
-        for (int i = 0; i < dag.nodes.length; i++) {
-            Node node = dag.nodes[i];
+        for (Node node : dag.nodes) {
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: We should be able to do something like this here:

```
dag.nodes.stream().
    filter(node -> node.node_type.equals(NodeType.ATTRIBUTE_NODE)).
    forEach(node -> LOGGER.info("{} = {}", node.attribute_name, CorbaUtils.getNodeValue(node.value)));
```

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 20]
```diff
@@ -270,65 +292,84 @@ public int getHitCount(Query query) throws Exception {
...
                 }
             }
         }
     }
 
-    public PackageElement[] order(ORB orb, DAG[] dags) throws Exception {
+    private URL getEncodedUrlFromString(String urlString) throws MalformedURLException {
+        URL url = new URL(urlString);
+        try {
+            URI uri = new URI(url.getProtocol(),
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: Could we validate that all the URL pieces are set instead of using the `URI` constructor? Or if that doesn't work, could we simply call `url.toURI()`? It just looks strange that we need to create a `URI` from the `URL` pieces and then convert that `URI` back to a `URL`.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 21]
```diff
@@ -270,65 +292,84 @@ public int getHitCount(Query query) throws Exception {
...
+    }
+
+    /**
+     * @param dag
+     * @return the first file name from the order response
+     * @throws Exception
+     */
+    public String testOrder(DAG dag) throws Exception {
+        PackageElement[] packageElements = order(dag);
+        return packageElements[0].files[0];
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: Could any one of those arrays be empty?

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 22]
```diff
@@ -408,32 +445,31 @@ public void retrieveProductFromDAG(DAG dag) throws MalformedURLException {
...
-                if (events != null) {
-                    for (Event event : events) {
-                        LOGGER.info("Event: {}\n Name: {}\n Desc: {}",
-                                event.event_type.value(),
-                                event.event_name,
-                                event.event_description);
-                    }
+            Event[] events = standingQueryMgr.get_event_descriptions();
+            if (events != null) {
+                for (Event event : events) {
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: Should be able to use `events.forEach()` here.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java (hunk 13 above)
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: +1

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 23]
```diff
@@ -492,180 +530,156 @@ public void testStandingQueryMgr(POA poa, Query query) throws Exception {
...
+        return get_related_files(getProductFromDag(dag));
+    }
+
+    public String[] get_related_files(Product product) throws Exception {
         if (productMgr != null) {
             LOGGER.info("Sending Get Related Files Request...");
 
-            FileLocation fileLocation = new FileLocation("", "", "", "", "");
-            NameValue[] properties = {new NameValue("", orb.create_any())};
+            FileLocation fileLocation = new FileLocation("user",
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: Could this be a constant?

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+    @AfterExam
+    public void stopHttpListener() {
+        if (server != null) {
+            server.stop();
+        }
+    }
+
+    @Before
+    public void startSampleNsiliClientFeature() throws Exception {
+        getServiceManager().startFeature(true, SAMPLE_NSILI_CLIENT_FEATURE_NAME);
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: Do we really need to start and stop this feature before and after every test? Couldn't we start it once when the container starts?

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    @After
+    public void stopSampleNsiliClientFeature() throws Exception {
+        getServiceManager().stopFeature(true, SAMPLE_NSILI_CLIENT_FEATURE_NAME);
+    }
+
+    @Test
+    public void testStandingQueryMgr() throws Exception {
+        SampleNsiliClient sampleNsiliClient = new SampleNsiliClient(Integer.parseInt(RESTITO_STUB_SERVER_PORT.getPort()), NSILI_ENDPOINT_IOR_URL.getUrl());
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: Should we extract a factory method to create the client or create the client in the `@Before` so we don't need to repeat this in all the tests?

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+
+        sampleNsiliClient.testStandingQueryMgr();
+
+        sampleNsiliClient.cleanup();
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: Can we move this in the `@After` method?

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 27]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    @Test
+    public void testCatalogMgrGetHitCount() throws Exception {
+        SampleNsiliClient sampleNsiliClient = new SampleNsiliClient(Integer.parseInt(RESTITO_STUB_SERVER_PORT.getPort()), NSILI_ENDPOINT_IOR_URL.getUrl());
+
+        assertThat(sampleNsiliClient.getHitCount(), is(1));
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: Not sure I understand how hit count should be one when we haven't done anything yet?

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 28]
```diff
@@ -0,0 +1,242 @@
...
+        final String expectedThumbnailFilename = ingestedProductId + "-THUMBNAIL.jpg";
+        assertThat(sampleNsiliClient.get_related_files(result), is(new String[]{expectedThumbnailFilename}));
+        assertStubServerPutRequest(expectedThumbnailFilename, "image/jpeg");
+
+        // OrderMgr
+        String orderResultFilename = sampleNsiliClient.testOrder(result);
+        // TODO test processing packageElements of the order response https://codice.atlassian.net/browse/CAL-192
+        assertStubServerPutRequest(orderResultFilename, "application/x-tar");
+
+        deleteMetacard(ingestedProductId);
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: `deleteMetacard` should be called in the `@After` method to make sure it always gets deleted.

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 29]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    private void startHttpListener() {
+        server = new StubServer(Integer.parseInt(RESTITO_STUB_SERVER_PORT.getPort())).run();
+        whenHttp(server).match(method(Method.PUT), Condition.startsWithUri(NSILI_FILE_URI_PATH))
+                .then(Action.success());
+    }
+
+    private void assertStubServerPutRequest(String expectedFilename, String expectedType) {
+        assertThat(server.getCalls().size(), is(1));
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: We should be able to use `verifyHttp()` instead of getting the calls and doing asserts on those. See https://github.com/mkotsur/restito#quick-example.

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java [hunk 30]
```diff
@@ -0,0 +1,242 @@
...
+                null);
+        Dictionary<String, Object> properties = new Hashtable<>();
+
+        properties.put("address",
+                DynamicUrl.SECURE_ROOT + HTTPS_PORT.getPort()
+                        + "/services/SecurityTokenService?wsdl");
+        stsClientConfig.update(properties);
+    }
+
+    public class NsiliSourceProperties extends HashMap<String, Object> {
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: We're trying to move away from extending `HashMap` in itests and instead have a private method that builds and returns a `Map` with the right values in it.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java (hunk 14 above)
Comment (emmberk) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/Client.java]: The `get_parameters` prints the attributes from the `productMgr.get_parameters` response

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java (hunk 21 above)
Comment (emmberk) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: This method is only called in `NsiliEndpointTest`, and we would expect neither array to be empty after having ingested one record.

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java (hunk 27 above)
Comment (emmberk) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: The `hitCount` is 1 because 1 record was ingested in the `@BeforeExam`

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java (hunk 21 above)
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: Can we always assume this will be the case? For instance, could some other tests or code use this client class in the future?

If that's the method's contract, i.e., the caller of this class needs to have ingested at least one thing, I would recommend adding that as a comment on the class or method.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java [hunk 31]
```diff
@@ -492,180 +530,156 @@ public void testStandingQueryMgr(POA poa, Query query) throws Exception {
...
-        if (productMgr != null) {
-            LOGGER.info("Sending Get Parameters Request to Server...");
+    public Product getProductFromDag(DAG dag) {
+        Product product = ProductHelper.extract(dag.nodes[0].value);
+        LOGGER.info("Product: {}", product.toString());
+        LOGGER.info("Product ID: {}", getProductIdFromDag(dag));
+        return product;
+    }
+
+    public DAG get_parameters(DAG dag) throws InvalidInputParameter, SystemFault, ProcessingFault {
```
Comment (lessarderic) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: Just noticed that we're using `_` in some method names. Can we get rid of those and use camel-case names instead?

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java (hunk 20 above)
Comment (emmberk) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: I updated this method to make it a little less clunky, but that was the simplest solution that I could find, but I'm open to suggestions.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java (hunk 21 above)
Comment (emmberk) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: I added javadoc comments to explain that this method is only used in the itest.

File: distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java (hunk 31 above)
Comment (emmberk) [inline distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java]: Updated.

I think the motivation behind the `_`s was to match the corresponding NSILI manager methods that will be called.

File: distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java (hunk 24 above)
Comment (emmberk) [inline distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/NsiliEndpointTest.java]: After wrestling the with itest dependencies and consulting @kcwire, I decided that making a new sample client and restarting that feature for each test method was the best solution for now. Please, let me know if you know of a better solution.

**Comment by troymohl** [comment]
Hero built this and ran the manual tests. Both pass.
